Lib/pickle.py

```diff
@@ -655,7 +655,24 @@ def save_reduce(self, func, args, state=None, listitems=None,
             self._batch_setitems(dictitems)
 
         if state is not None:
-            save(state)
+            msg = "state must be either a dict or a tuple of length 2"
+            if state_setter is None:
+                save(state)
+
+            elif isinstance(state, dict):
```

Comment:
I wonder if we wouldn't be better off with a separate opcode instead of reusing BUILD with a new argument syntax. This would put this change in the realm of pickle protocol 5 (making it harder for you to develop and test until PEP 574 is merged ;-)). To be discussed on python-dev? @ncoghlan

Reply:
Actually @pierreglaser suggested IRL to use a nested reduce when we want to use a custom state_setter. This way we would not pervert the semantics of `BUILD` and the resulting pickle would be loadable in older Python versions.

Reply:
Yeah, I implemented this new method. It is fully backward compatible, but kind of hacky - I'll let you see by yourselves: https://github.com/pierreglaser/cpython/pull/2